### ISSUE TITLE
Tesseract 4.0.0 crashed on Intel I5-8400 CPU with Debian 9.6.0 amd64 (SSE/AVX/AVX2)

### ISSUE THREAD

**Issue by edilinux** [tesseract-ocr/tesseract#2064]
### Environment

* **Tesseract Version**:  4.0.0 Release
* **Commit Number**:  51316994ccae0b48692d547030f26c0969308214
* **Platform**: Debian 9.6.0 amd64

### Current Behavior: Tesseract 4.0.0 crashed on Itel I5-8400 CPU with Debian 9.6.0 amd64 (SSE/AVX/AVX2).
I compiled the tesseract 4.0 on Itel I5-8400 CPU with Debian 9.6.0 amd64.
tesseract --version output this:
tesseract 4.0.0
leptonica-1.74.2
 libjpeg 6b (libjpeg-turbo 1.5.1) : libpng 1.6.28 : libtiff 4.0.8 : zlib 1.2.8
 Found AVX2
 Found AVX
 Found SSE

When I call tesseract several times, crash happens and PC is reboot.

I have a Intel G4650 CPU and this CPU not suport AVX2 / AVX and everything works fine! Never crash happens!
How to make tesseract work fine on Intel I5-8400 with AVX/AVX2/SSE.

**Comment by amitdo** [comment]
>When I call tesseract several times, crash happens and PC is reboot.

From the C/C++ API? Paste the code you are using.

Do you run this code inside a VM?

Does running tesseract command line tool work fine?

Did you test rc/beta versions of tesseract 4.0.0?

**Comment by edilinux** [comment]
"From the C/C++ API? Paste the code you are using."
I'm not use the API, I use command line of linux:
tesseract img.jpeg img.jpeg.txt -l fra+eng hocr;
tesseract img2.jpeg img2.jpeg.txt -l fra+eng hocr;
....

"Do you run this code inside a VM?"
Yes, and works fine! because virtual box not expose AVX/AVX2/SSE instructions.
But, when I run on real SO with I5 8400 CPU tesseract crashed and system reboot.
On real SO with G4650 CPU works fine.

"Does running tesseract command line tool work fine?"
Not, when I run several times this line:
tesseract img.jpeg img.jpeg.txt -l fra+eng hocr;
The system reboot.

"Did you test rc/beta versions of tesseract 4.0.0?"
not yet, just tested the version 4.0.0 release.
What version you recommend to test?

**Comment by stweil** [comment]
As you compile Tesseract yourself, you could locally modify [the code](https://github.com/tesseract-ocr/tesseract/blob/master/src/arch/simddetect.cpp#L58) to pretend that your CPU does not support AVX2 by setting `avx2_available_ = false`. Does that help? If yes, we must examine that closer. If not, you also have to set `avx_available_ = false`.

But there is another more fundamental problem: user code like Tesseract should never cause a Linux system crash. Can you reproduce the crash with the official [Debian tesseract-ocr package 4.0.0](https://packages.debian.org/search?keywords=tesseract-ocr&searchon=names&exact=1&suite=all&section=all) (from stretch-backports or newer)? Then I'd report an issue to Debian.

**Comment by stweil** [comment]
I assume that Tesseract does not crash when you just run `tesseract -v` several times, but crashes while doing real OCR. Does it crash with any image? How often do you have to run it until it crashes with a small image? Does it crash faster when doing OCR of a large image? Do you get any crash information, either a short message on the screen or messages in `/var/log/syslog`?

Recently I noticed that the Tesseract code uses AVX(2) with a wrong alignment of data (4-byte aligned instead of 8-byte aligned). That might explain performance problems (especially observed on Windows, but also on Linux). I could not examine it closer up to now, but perhaps such misaligned data causes a kernel trap causing a system reboot in rare cases (that's pure speculation up to now).

**Comment by edilinux** [comment]
"I assume that Tesseract does not crash when you just run tesseract -v several times, but crashes while doing real OCR. Does it crash with any image? How often do you have to run it until it crashes with a small image? Does it crash faster when doing OCR of a large image? Do you get any crash information, either a short message on the screen or messages in /var/log/syslog?"
I have a script with: 
tesseract img.jpeg img.jpeg.txt -l fra+eng hocr;
There are 5.000 imagens to process on my script.
when the execution is in the fifth line, the system reboot, no error show me, no log in the /var/log/syslog.
The system simply restarts.

"Recently I noticed that the Tesseract code uses AVX(2) with a wrong alignment of data (4-byte aligned instead of 8-byte aligned). That might explain performance problems (especially observed on Windows, but also on Linux). I could not examine it closer up to now, but perhaps such misaligned data causes a kernel trap causing a system reboot in rare cases (that's pure speculation up to now)."
Okay, I understand, thanks for the support.
I will try to compile in ubuntu and with other versions and return to comment here.

**Comment by stweil** [comment]
Using AVX instead of AVX2 won't cost much performance, and even SSE still has acceptable performance. An unexpected system reboot will cost much more, I expect.

On Windows @noahmetzger recently reported that not using SSE or AVX is faster than using AVX.

**Comment by edilinux** [comment]
"Using AVX instead of AVX2 won't cost much performance, and even SSE still has acceptable performance. An unexpected system reboot will cost much more, I expect."
Can I disable AVX2 on tesseract-ocr package? or I'll have to compile again.

**Comment by stweil** [comment]
>  Can I disable AVX2 on tesseract-ocr package? or I'll have to compile again.

No, it cannot be disabled by the user (that's a desired feature), so you have to compile again.

**Comment by stweil** [comment]
> when the execution is in the fifth line, the system reboot,

What means "fifth line" here? Does the system reboot while Tesseract is processing the 5th image? Would it also reboot immediately if you start with that image? If yes, then that image is special, and it would be good if you could share it somehow, so it is possible to examine it closer.

**Comment by edilinux** [comment]
yes, it's the fifth image.
I also thought it was the image, but with tesseract 3 it works.
And without AVX2 too.

**Comment by stweil** [comment]
Can you attach the image to the issue report? Or send it to me per e-mail if the content is sensitive?

**Comment by edilinux** [comment]
Yes I can, I am currently not having access to the image, but I am sending it to you as soon as possible.
Thanks a lot!

**Comment by edilinux** [comment]
More news,
The problem in the image is discarded, because I tested with other images and touched the same problem.
I also realized that if I use tesseract in only 1 image (with the image that failed) everything works.

I disabled AVX2 (avx2_available_= false) and it worked, but I'm not sure if that's the problem.
I tested the tesseract with AVX2:
* On the i5-7400 + debian and it worked.
* On the I5-8400 but I substituted debian for ubuntu server and it also worked.
Apparently it is not a problem of tesseract but of debian.
And for me it's okay to use ubuntu, it's extremely fast (tesseract run 40% more fast).

**Comment by stweil** [comment]
Thank you for reporting those new test results. That's very interesting and adds more hints that there might be a problem with Tesseract's implementation of the dot product calculation on AVX2. There is only a AVX2 implementation for integer models, so I assume the reported problem can only be seen with models from tessdata_fast (the models provided by the major Linux distributions), but not with models from tessdata_best or tessdata.

You could try to install a newer Linux kernel on Debian (for example from backports) and test whether that improves the performance and fixes the crashes. Which exact kernel versions do you currently use on Debian / on Ubuntu?

How does Tesseract performance with AVX (AVX2 disabled) compare to Tesseract with AVX2 on Ubuntu? Do you get performance differences with AVX on Debian compared to Ubuntu?

**Comment by amitdo** [comment]
I also realized that if I use tesseract in only 1 image (with the image that failed) everything works.

I don't understand what you mean here. Can you clarify?

**Comment by stweil** [comment]
I now have run a benchmark using [tessdata_best/script/Latin.traineddata]() on one of our Debian GNU Linux servers with different Linux kernels:

linux-image-4.9.0-8-amd64 4.9.130-2:

    real    3m52.251s
    user    12m16.976s
    sys     0m7.152s

linux-image-4.18.0-0.bpo.1-amd64 4.18.6-1~bpo9+1:

    real    2m49.634s
    user    8m33.314s
    sys     0m4.478s

So the newer kernel greatly increases the speed of Tesseract!

**Comment by edilinux** [comment]
bad news,
Ubuntu 18.04.1 crashed too.
For now, please disregard the tests I've done.
Because what is making a difference in this bug is the amount of images processed.
In Debian after the fifth image the system crashes and in Ubuntu after 1200 images the system crashes.
I will redo all the tests I did, this time with 10000 images or more.
How to run tesseract 4 many times in parallel? In tesseract 3 I used parallel (gnu.org/software/parallel) but it does not work with tesseract 4.
How to do this? Does tesseract 4 work in an HPC cluster?

**Comment by edilinux** [comment]
"So the newer kernel greatly increases the speed of Tesseract!"
How was your test? Did you use AVX / AVX2? how many images have been processed?
Did you use tesseract in parallel?

**Comment by amitdo** [comment]
The difference in speed might be related to spectra and meltdown patches.

**Comment by stweil** [comment]
> The difference in speed might be related to spectra and meltdown patches.

Yes, I think so, too. Obviously newer kernels have a better implementation to handle those CPU bugs. Tests without any such patches (they can be disabled via kernel parameters) are still to be done.

> How was your test?

I have run OCR for a single [image](https://digi.bib.uni-mannheim.de/~stweil/tesseract/0604.png) using `time tesseract 0604.png output -l script/Latin`, so Tesseract used its internal multithreading.

For your use case with thousands of images, it is much better to run Tesseract single-threaded (`export OMP_THREAD_LIMIT=1` in the environment) and use _parallel_ or other scripts to run as many Tesseract instances as there are CPU cores.

**Comment by stweil** [comment]
@edilinux, did you run Tesseract as root or as a normal user when it crashed your system? I'd like to reproduce it.

**Comment by edilinux** [comment]
> @edilinux, did you run Tesseract as root or as a normal user when it crashed your system? I'd like to reproduce it.

You should process more than 2mil images for the system to crash.
I'm redoing the tests, but I believe if your CPU does not have AVX / AVX2 then the system will not crash.

**Comment by stweil** [comment]
My performance test was done on a server with AVX2 support. I'll repeat it with smaller images running in a loop. Did you run several instances of Tesseract in parallel when you got a system reboot?

**Comment by stweil** [comment]
I have now run OCR for [phototest.tif](https://github.com/tesseract-ocr/test/blob/master/testing/phototest.tif) 60000 times on the server with AVX2 without a crash:

    # using eng.traineddata from tessdata_fast
    tesseract test/testing/phototest.tif /tmp/out -l eng

**Comment by edilinux** [comment]
@stweil @amitdo  Thanks a lot for your help!
I am convinced that this is some problem with my equipment / installation.
I'm still testing but I have not had problems so far with an I5 7400 CPU + Debian + 4.9.0-8-amd64 + AVX + AVX2.

**Comment by edilinux** [comment]
I don't believe, I found the problem.
One of my partitions was with btrfs and I can not explain why the system crashes when using tesseract.
It's strange because the images were in a partition with Ext4.
@stweil @amitdo sorry for bothering and wasting your time, all the time I thought it was something with AVX / AVX2 because on the CPU that was working that was the only difference I had noticed.
And one more time thanks a lot for your help!

**Comment by stweil** [comment]
So the result of the OCR was written to a BTRFS partition? That still should not be a reason for a kernel reboot.

But your issue report helped me very much, as I was not aware of the huge impact of the Linux kernel version on the performance of Tesseract OCR.

**Comment by edilinux** [comment]
Yes, was written to a btrfs partition, and the /root directory was on btrfs partition too.
I know it's strange, but after changing btrfs for ext4 everything is working fine.
the CPU I5 8400 is on for more than 20 hours and has not crashed.

> But your issue report helped me very much, as I was not aware of the huge impact of the Linux kernel version on the performance of Tesseract OCR.

Yes, the difference in performance is impressive, tesseract 4 + newer kernel is extremely powerful and fast!
My setup for better performance:
- Debian 9 + Ext4 + linux-image-4.17.0-0.bpo.3-amd64
- tesseract 4.0.0
 leptonica-1.76.0
  libgif 5.1.4 : libjpeg 6b (libjpeg-turbo 1.5.1) : libpng 1.6.28 : libtiff 4.0.8 : zlib 1.2.8 : libwebp 0.5.2 : libopenjp2 2.3.0
 Found AVX2
 Found AVX
 Found SSE

Best regards

**Comment by zdenop** [comment]
Seems like those who search for any speed improvement should test different
linux kernels:

https://www.phoronix.com/scan.php?page=article&item=linux-410-420&num=1

Zdenko


št 22. 11. 2018 o 8:19 Stefan Weil <notifications@github.com> napísal(a):

**Comment by zdenop** [comment]
@edilinux : if the issue is solved, please close it.

**Comment by hieonn** [comment]
@stweil Thanks for your effort, I had same problem on my environment, and stuck to this problem several days. I don't have same problem any more by disabling AVX features like you advised. The environment of mine for now is like followings.

Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
Ubuntu 16.04 LTS 
tesseract 4.1.0-rc1
 leptonica-1.77.0
  libjpeg 8d (libjpeg-turbo 1.4.2) : libpng 1.2.54 : libtiff 4.0.6 : zlib 1.2.8 : libopenjp2 2.1.2
 Found SSE

Thanks!

**Comment by edilinux** [comment]
> @stweil Thanks for your effort, I had same problem on my environment, and stuck to this problem several days. I don't have same problem any more by disabling AVX features like you advised. The environment of mine for now is like followings.
> 
> Intel(R) Core(TM) i7-7700 CPU @ 3.60GHz
> Ubuntu 16.04 LTS
> tesseract 4.1.0-rc1
> leptonica-1.77.0
> libjpeg 8d (libjpeg-turbo 1.4.2) : libpng 1.2.54 : libtiff 4.0.6 : zlib 1.2.8 : libopenjp2 2.1.2
> Found SSE
> 
> Thanks!
The solution for me.
https://github.com/tesseract-ocr/tesseract/issues/2064#issuecomment-441012762

**Comment by hieonn** [comment]
@edilinux Thanks for sharing your information. You put all OCR stuffs in btrfs and the problem has gone? I will give it a try if my understanding is right. Please confirm it. Thanks!

**Comment by jeodai** [comment]
I have the similar crush problem.

tesseract --version
  tesseract 4.0.0
  leptonica-1.74.1  libgif 5.1.4 : libjpeg 6b (libjpeg-turbo 1.5.1) : libpng 1.6.28 : libtiff 4.0.8 : zlib 1.2.8 : 
  libwebp 0.5.2 : libopenjp2 2.1.2
 Found AVX2
 Found AVX
 Found SSE

And I tested both on centos 7.5 and debian 9.8

**Comment by stweil** [comment]
@jeodai, please give more details. Are you running on real hardware or in a virtual machine? Does Tesseract crash with any image? Does your PC reboot?

**Comment by mikegerber** [comment]
@stweil I only tested this with one picture which I can send to you privately (it's from a data set and I cannot publish the image.) If it helps I can also test with other pictures but for now I'm happy that my computer doesn't crash anymore.

**Comment by stweil** [comment]
I expect that it will crash with any picture. Can you confirm that? Are you running on real hardware or in a virtual machine? Does your PC reboot?

**Comment by mikegerber** [comment]
It also crashes with http://www.mattmahoney.net/ocr/stock_gs200.jpg on the second attempt. I use real hardware and the computer hard resets (goes off and on again) with no error message visible.

**Comment by mikegerber** [comment]
I also tested this in a chroot running the GRML 2018.12 kernel (Debian based rescue system) with the same hard reset effect.

**Comment by stweil** [comment]
Instead of modifying the code, you can also tell Tesseract to avoid the AXV2 code: `tesseract IMAGE OUTPUTBASE [...] -c dotproduct=sse`.

What happens if you run your test with gdb supervision? `gdb --args tesseract [...]`

**Comment by mikegerber** [comment]
With `-c dotproduct=sse` I do not get the hard resets. If I run it using `gdb` and with the AVX2 code, the system also hard resets after a few tries.

**Comment by mikegerber** [comment]
FWIW, I run everything on ext4.

**Comment by stweil** [comment]
I don't think that it is related to the filesystem.

Summary of the currently known facts:

We now have at least two cases with Intel Core i5-8400, a CPU which claims to support AVX2 but not only crashes when running Tesseract with AVX2 code but even reboots Linux.

That could be a bug in the Intel microcode. @edilinux and @mikegerber, which microcode package do you have installed? Is it current?

Using SSE 4.2 instead of AVX2 by calling Tesseract with `-c dotproduct=sse` is a working workaround.

**Comment by mikegerber** [comment]
~~~
~ dmesg | grep -i microcode
[    0.000000] microcode: microcode updated early to revision 0xb4, date = 2019-04-01
[    1.258506] microcode: sig=0x906ea, pf=0x2, revision=0xb4
[    1.258691] microcode: Microcode Update Driver: v2.2.
~~~

If I interpret this message correctly, the microcode seems to be up-to-date.

**Comment by stweil** [comment]
My last idea is running Tesseract in a pure Linux console (Alt-Ctrl-F1, no X environment) and hope that the Kernel shows a last piece of information on that console before it reboots. Maybe you need a video camera to film the very short moment when this information is on the display.

**Comment by amitdo** [comment]
>On Windows @noahmetzger recently reported that not using SSE or AVX is faster than using AVX.


Is there a separate issue for this?

**Comment by stweil** [comment]
I am not sure whether this is still true and will ask @noahmetzger on Monday or test it myself.

**Comment by mikegerber** [comment]
Apparently I do not have this problem anymore.

* Same microcode version [as above](https://github.com/tesseract-ocr/tesseract/issues/2064#issuecomment-496849042)
* Using Tesseract 4.1.0
* Kernel 5.2.11 now

🤔

~~~
% bin/tesseract --version
tesseract 4.1.0
 leptonica-1.78.0
  libgif 5.1.4 : libjpeg 6b (libjpeg-turbo 2.0.2) : libpng 1.6.34 : libtiff 4.0.10 : zlib 1.2.11 : libwebp 1.0.3
 Found AVX2
 Found AVX
 Found SSE
 Found libarchive 3.3.3 zlib/1.2.11 liblzma/5.2.4 bz2lib/1.0.6 liblz4/1.8.3 libzstd/1.3.8
~~~

**Comment by qbeer** [comment]
Same issue here, the system is running tesseract fine with the `-c dotproduct=sse` flag but hard resets every time I run it without it. Still no solution, it hugely inconvenient.

**Comment by zdenop** [comment]
Do you use the latest version?

**Comment by qbeer** [comment]
It happened several times on `version 4` as well, after that I downloaded `version 5`. I use a `Ubuntu 20.04 LTS`.

```bash
tesseract 5.0.0-alpha-20201231-86-g95b9
 leptonica-1.79.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 2.0.3) : libpng 1.6.37 : libtiff 4.1.0 : zlib 1.2.11 : libwebp 0.6.1 : libopenjp2 2.3.1
 Found AVX2
 Found AVX
 Found FMA
 Found SSE
 Found OpenMP 201511
 Found libarchive 3.4.0 zlib/1.2.11 liblzma/5.2.4 bz2lib/1.0.8 liblz4/1.9.2 libzstd/1.4.4
```

**Comment by zdenop** [comment]
Did you build it with autotools or cmake?

**Comment by qbeer** [comment]
Got it from the `apt` repository, but currently I am building it locally as well since I am seeing gibberish results even when it runs on a standard document (white background + black text).

**Comment by sebastian-sohr** [comment]
For me, tesseract 4.1.1 crashes with german trained data on a ubuntu 20.04 system (cpu: Intel® Core™ i5-6600 CPU @ 3.30GHz × 4):
Command:
`tesseract ./csm_Anmerkung_2021-03-14_102110_8bfe96c78b.jpg stdout -l deu`

The english version doesn't crash (put produces pure result because the image requires preprocessing, for sure).

Tesseract version:

```
$ tesseract -v
tesseract 4.1.1
 leptonica-1.79.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 2.0.3) : libpng 1.6.37 : libtiff 4.1.0 : zlib 1.2.11 : libwebp 0.6.1 : libopenjp2 2.3.1
 Found AVX2
 Found AVX
 Found FMA
 Found SSE
 Found libarchive 3.4.0 zlib/1.2.11 liblzma/5.2.4 bz2lib/1.0.8 liblz4/1.9.2 libzstd/1.4.4
```
 System:
 
` Linux sohr-1 5.8.0-45-generic #51~20.04.1-Ubuntu SMP Tue Feb 23 13:46:31 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`


![csm_Anmerkung_2021-03-14_102110_8bfe96c78b](https://user-images.githubusercontent.com/53340074/112002525-44895f80-8b20-11eb-95d6-6a37f1b5c84b.jpg)

**Comment by stweil** [comment]
I just tried that on Ubuntu 20.10 and had no crash. Can you debug the crash with GDB?

**Comment by sebastian-sohr** [comment]
I'm no specialist: Can you give me a command?

**Comment by stweil** [comment]
Run `gdb --args tesseract ./csm_Anmerkung_2021-03-14_102110_8bfe96c78b.jpg stdout -l deu` from the shell command line to load the program with the GDB debugger.

Then enter `run` in the debugger. As soon as it has crashed, enter `info stack`. Post the output here.

**Comment by sebastian-sohr** [comment]
`$ gdb --args tesseract ./csm_Anmerkung_2021-03-14_102110_8bfe96c78b.jpg stdout -l deu
GNU gdb (Ubuntu 9.2-0ubuntu1~20.04) 9.2
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from tesseract...
(No debugging symbols found in tesseract)
(gdb) run
Starting program: /usr/bin/tesseract ./csm_Anmerkung_2021-03-14_102110_8bfe96c78b.jpg stdout -l deu
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff4b77700 (LWP 5349)]
[New Thread 0x7ffff4376700 (LWP 5350)]
[New Thread 0x7ffff3b75700 (LWP 5351)]
Anzahl Infizierter D davon aufgrund ] _ davon schwere

letzten 7 Tage* Patienten stationär COVID-19 Verstorbene’

Verläufe? Anzahl

Sanl Genesener“
Landkreise / kreisfreie aktueller | in den letzte 7 | Gesamt- | letzte7 | Gesamt- |letzte 7 |Gesamtz| letzte 7

Tage® | zahl Tage® zahl Tage | ahl | Tage®

Meldung Anzahl |pro 100.000|Gesamt-zahl

Städte Stand |letzten 24 h
‚vom

27 167, z 186 o 10 a

42 23

152
Nordhausen‘ 46
79

223

96
2808 2818 38 399 145
4904 4928 157 474 210

Weimarer 2567 2567 50 162 51
Eisenach 1088 1094 39 49
Erfurt 5553 5587 196 481 198

Jena 3019 3038 19 74 213 59

Weimar. 1828 1835 7 33 137 71 2 93
82760 632 188 3064

Weiße Färbung: Inzidenz 0- <50, heile Färbung: Inzidenz 50- < 100, mittlere Färbung: Inzidenz 100- <200, dunkle Färbung: Inzidenz 200- <400, stark dunkle Färbung: Inzidenz 2400

* Vortag minus 6 Tage 0:00 Uhr bis Vortag 24:00 Uhr

? schwerer Verlauf definiert als Aufnahme auf eine Intensivstation

° an COVID-19 bzw. mit SARS-CoV-2 verstorben

“ geschätzt entsprechend dem Algorithmus des RKI vom 08.04.2020. Berücksichtigt werden Hospitalisierungsstatus, Status Verstorben, Erkrankungsbeginn und ggf. Meldedatum.

*Eingang der Meldung am TLV

© Es ist von Nachmeldungen auszugehen.

7 Datenkorrektur durch das Gesundheitsamt

 

[Thread 0x7ffff4376700 (LWP 5350) exited]
[Thread 0x7ffff4b77700 (LWP 5349) exited]
[Thread 0x7ffff4b9d7c0 (LWP 5345) exited]
[Inferior 1 (process 5345) exited normally]
(gdb) 
`

**Comment by stweil** [comment]
So `tesseract` did not crash when running with GDB. Now try `ulimit -c unlimited` before running `tesseract` (without GDB). This will create a so called core dump file (typically named `core`or similar) as soon as the program crashed. Then try `gdb tesseract core*` from the command line and ìnfo stack` in gdb.

**Comment by sebastian-sohr** [comment]
After calling `ulimit -c unlimited` running 

`tesseract ./csm_Anmerkung_2021-03-14_102110_8bfe96c78b.jpg stdout -l deu `

causes no crash any more.

**Comment by stweil** [comment]
That's strange. But fine, without crash we cannot examine the crash. If it happens again, you can try the above commands to get more information.

**Comment by sebastian-sohr** [comment]
I already tried again. But now it crashes also after calling ulimit -c unlimited (and checking with ulimit -a).
But it no core is created.

**Comment by stweil** [comment]
Are you sure? Look for a new file – maybe it got a different name. I never heard that Linux did not create a core file for a crash after running `ulimit -c unlimited` in the same shell. Of course there must be sufficient disk space available for writing the file.

**Comment by stweil** [comment]
What is the exact crash message?

**Comment by sebastian-sohr** [comment]
Nothing: It simply reboots. I also didn't find anything in syslog.
I will try running it in a docker container:
I want to see whether the host system reboots, or the container crashes.

**Comment by stweil** [comment]
So running Tesseract reboots your Linux system? That's even more strange. In that case I'd try reproducing the issue in a Linux console (no X Windows) and make a smartphone video to see the error message which is reported by the kernel.

**Comment by mikegerber** [comment]
We should consider if this could be a power supply problem: AVX2 does cause power surges/requires more power and it is conceivable that a underpowered/bad power supply causes exactly this kind of error: The PC simply hard resets (= turns off).

@sebastian-sohr Did you try disabling AVX code using `-c dotproduct=sse`?

**Comment by sebastian-sohr** [comment]
@sebastian-sohr Did you try disabling AVX code using -c dotproduct=sse?

Yes, but I got only binary  (garbage) data as output.

**Comment by mikegerber** [comment]
>>  @sebastian-sohr Did you try disabling AVX code using -c dotproduct=sse?
> Yes, but I got only binary (garbage) data as output.

Strange, I suspect that this is another problem unrelated to the crashing.

**Comment by stweil** [comment]
Nevertheless it did not crash with `-c dotproduct=sse`, so that could confirm that the reset is related to AVX and maybe problems with the power supply.

**Comment by mikegerber** [comment]
> So running Tesseract reboots your Linux system? That's even more strange. In that case I'd try reproducing the issue in a Linux console (no X Windows) and make a smartphone video to see the error message which is reported by the kernel.

If it's anything like the problem I had (and where you also asked), then *there is no kernel error*, the system simply hard resets, turns off. No message, no debugging, just turning off. That's why I suspect a power supply or other hardware problem, triggered by the unusual AVX2 CPU load.

**Comment by sebastian-sohr** [comment]
> That's why I suspect a power supply or other hardware problem, triggered by the unusual AVX2 CPU load.

Hm: I get no reboot when running the command in a docker container:

```
tesseract 4.1.1
 leptonica-1.79.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 2.0.3) : libpng 1.6.37 : libtiff 4.1.0 : zlib 1.2.11 : libwebp 0.6.1 : libopenjp2 2.3.1
 Found AVX2
 Found AVX
 Found FMA
 Found SSE
 Found libarchive 3.4.0 zlib/1.2.11 liblzma/5.2.4 bz2lib/1.0.8 liblz4/1.9.2 libzstd/1.4.4

```
```
root@3bc932cf0f16:/# tesseract ./csm_Anmerkung_2021-03-14_102110_8bfe96c78b.jpg stdout -l deu
Anzahl Infizierter Naunnelertenst davonaufgrund | davonschwere

letzten 7 Tage" Patienten stationär CoVID-19 Verstorbene

Verläufe? Anzahl

Stand Genesener®
Landkreise / kreisfreie aktueller | inden letzte 7 | Gesamt- | letzte? | Gesamt- | letzte 7 |Gesamtz letzte 7
...
```

**Comment by stweil** [comment]
Maybe the docker container limits the number of threads? Try `OMP_THREAD_LIMIT=1 tesseract ...` outside of Docker. Then only one AVX unit will be used instead of up to four, so the power consumption should be reduced. If that works, you can try raising the limit.

**Comment by sebastian-sohr** [comment]
Yes, it works with `export OMP_THREAD_LIMIT=1`.
But I can't see whether the number of threads is restricted in the docker container.

But I see that the results are different - so somehow the systems behave differently:

Host:
```
$ tesseract ./csm_Anmerkung_2021-03-14_102110_8bfe96c78b.jpg stdout -l deu
Anzahl Infizierter D davon aufgrund ] _ davon schwere

```
Docker:

```
root@3bc932cf0f16:/# tesseract ./csm_Anmerkung_2021-03-14_102110_8bfe96c78b.jpg stdout -l deu
Anzahl Infizierter Naunnelertenst davonaufgrund | davonschwere
```
Strange, but, probably, not everything can be explained ...

**Comment by mohamedation** [comment]
i am also facing this issue with hard reset when doing multiple images. using `-c dotproduct=sse` does make the command complete without the hard reset, but with garbage as @sebastian-sohr  said.

**Comment by stweil** [comment]
@mohamedation, does it work if you set `export OMP_THREAD_LIMIT=1` before running `tesseract`?

**Comment by mohamedation** [comment]
> @mohamedation, does it work if you set `export OMP_THREAD_LIMIT=1` before running `tesseract`?

@stweil yes it does.

**Comment by stweil** [comment]
Then the hard reset might also be caused by problems with the electrical power supply. Your computer is able to run a single AVX2 unit, but not 4 of them. You can try whether it works at least with 2 AVX2 threads by setting `export OMP_THREAD_LIMIT=2`.

**Comment by logological** [comment]
I am experiencing the same problem with Tesseract 4.1.1 running on openSUSE Tumbleweed with Linux 6.1.1.  My system has an Intel Core i7-8700 @ 3.20 GHz.  When I try to use Tesseract to OCR certain files, my entire computer reboots.  The problem is consistently reproducible.

Adding `-c dotproduct=sse` to the command line works around the problem.  Strangely, using `export OMP_THREAD_LIMIT=1` before running Tesseract causes Tesseract to emit the following error message:

```
Error opening data file /app/share/tessdata/deu.traineddata
Please make sure the TESSDATA_PREFIX environment variable is set to your "tessdata" directory.
Failed loading language 'deu'
Missing = in configvar assignment
```

I have another computer with an identical software configuration (exact same versions of Tumbleweed, the Linux kernel, and Tesseract) but with an Intel Core i7-7700K CPU, and it does not exhibit the problem.

Since several other people here have reported the same hard reset behaviour, I'm skeptical that this is a purely electrical issue.  (Can it really be that _all_ of our computers have failing or underpowered power supplies?)  Could this be a bug in the Linux kernel, or in the CPU's microcode, or some design flaw with the CPU itself?  Should this be reported upstream anywhere, and if so, where?

**Comment by zdenop** [comment]
Tesseract version 4 is outdated and unsupported. Please use the recent version of the tesseract (5.3)

**Comment by amitdo** [comment]
There is probably no bug in our code that causes this issue.This looks like a bug in the CPU power management. 

i think you should seek Intel's support.

Somebody already reported about this issue years ago:

https://community.intel.com/t5/Processors/tesseract4-0-crash-when-I-use-Intel-i7-8700K-CPU-with-Ubuntu-18/td-p/609019

**Comment by qyz** [comment]
Hi,

just for the record, I'm hitting the same bug on Debian 12 with Debian's tesseract-5.3.0-2.

I have a simple pipeline for archiving personal documents using `sane`/`scanimage`, `imagemagick` and `tesseract`.

The the system and the archive process had worked absolutely stable for more than 5 years from Debian Stretch through Bookworm on a Celeron G3930 processor.  A few weeks ago I upgraded the processor to a i7-7700K (and added more RAM) and now the system hard-resets in about one of every four tesseract runs. Otherwise the system runs stable, there were no hard resets without tesseract involvement.

Headless system,
Default Debian 12 Kernel 6.1.0-11
Motherboard Asus PRIME B250M-C
i7-7700K, 48 GB RAM,
1 TB NVMe SSD (ext4/root), 2 x 8TB SATA HDD (zfs/data), 
600W Power Supply.

I've set `OMP_THREAD_LIMIT=1` and call tesseract with `-c dotproduct=sse`. No resets so far in about 12 runs.

Cheers,
qyz
